mypy/fastparse.py

```diff
@@ -403,15 +404,36 @@ def visit_Delete(self, n: ast35.Delete) -> Node:
         else:
             return DelStmt(self.visit(n.targets[0]))
 
-    # Assign(expr* targets, expr value, string? type_comment)
+    # Assign(expr* targets, expr? value, expr? type_comment)
     @with_line
     def visit_Assign(self, n: ast35.Assign) -> Node:
         typ = None
+        if hasattr(n, 'new_syntax') and n.new_syntax == 1:  # type: ignore
+            n_synt = True
+        else:
+            n_synt = False
+        if n_synt and self.pyversion < (3, 6):
+            raise TypeCommentParseError('Variable annotation syntax is only '
+                                        'suppoted in Python 3.6, use type '
+                                        'comment instead', n.lineno)
         if n.type_comment:
-            typ = parse_type_comment(n.type_comment, n.lineno)
-
-        return AssignmentStmt(self.visit_list(n.targets),
-                              self.visit(n.value),
+            if isinstance(n.type_comment, str):  # backward-compatibility typed_ast
+                typ = parse_type_comment(n.type_comment, n.lineno)
+            elif isinstance(n.type_comment, ast35.Str):
+                typ = parse_type_comment(n.type_comment.s, n.lineno)
+            else:
+                typ = TypeConverter(line=n.lineno).visit(n.type_comment)
+        if n.value is None:  # always allow 'x: int'
+            rvalue = TempNode(AnyType())  # type: Node
+        else:
+            rvalue = self.visit(n.value)
+        lvalues = self.visit_list(n.targets)
+        # protect 'x, y = None, None # type: int, str' etc.
+        if not n_synt and typ is not None and experiments.STRICT_OPTIONAL:
```

Comment:
This understanding should be part of a later stage rather than in the parser.  It's better if (as much as possible) the parser faithfully reports the AST of the actual code.